Add SeekablePathStream for java.nio.file.Path.

tomwhite opened the pull request.

zong0jie commented:
Good works. With this PR I think It can manipulate hdfs-bam-files to some extent. Next time maybe I can just ask here may need a function instead of make a PR, and you will give good solutions :)

tomwhite commented:
Thanks @zong0jie. Please continue to submit PRs :)

The test failure seems to be an unrelated transient failure (all tests pass locally for me), but I don't seem to have permission to rerun it.

droazen commented:
@tomwhite I restarted the travis build just now.

tomwhite commented:
New test failure in RelativeIso8601DateTest is again unrelated.

tomwhite commented:
@droazen can you review this please when you have time? The test failure is unrelated.

droazen commented:
@cmnbroad can you take a look at this when you have time?

cmnbroad commented on src/java/htsjdk/samtools/SamReaderFactory.java:
```diff
@@ -56,6 +57,13 @@
     
     abstract public SamReader open(final File file);
 
+    public SamReader open(final Path path) {
```
We should probably add a test for this, since I don't think its covered elsewhere.

cmnbroad commented on src/java/htsjdk/samtools/SamInputResource.java:
```diff
@@ -192,6 +209,62 @@ public File asFile() {
 
     @Override
+    public Path asPath() {
+        return null;
```
Should return a Path for the file here.

cmnbroad commented on src/java/htsjdk/samtools/SamInputResource.java:
```diff
@@ -234,6 +307,11 @@ public File asFile() {
     @Override
+    public Path asPath() {
+        return null;
+    }
```
We should probably try to return a Path here as well ?

cmnbroad commented on src/java/htsjdk/samtools/seekablestream/SeekablePathStream.java:
```diff
@@ -0,0 +1,123 @@
+                sfs.close();
+            } catch (IOException e) {
+                //TODO
+                //log.error("Error closing SeekableFileStream", e);
```
Seems like we should re-throw, or at the very least log.

cmnbroad commented:
Sorry about the delay in getting to this @tomwhite. Looks pretty good; a few minor comments, mostly about filling out the SamInputResource conversion methods.

tomwhite commented:
Thanks for the review @cmnbroad! I've addressed all your review comments now.

akiezun commented:
@droazen ping

tomwhite commented:
@droazen could you take a look at this when you have time.

yfarjoun commented on src/java/htsjdk/samtools/SamFiles.java:
```diff
@@ -19,7 +21,19 @@
+     *
+     * @return The index for the provided SAM, or null if one was not found.
+     */
+    public static Path findIndex(final Path samFile) {
```
variable name samPath?

tomwhite commented:
Addressed @yfarjoun's feedback about naming (thanks!), and rebased. This should be ready for merge now.

droazen commented:
@tomwhite Ahem, well actually this branch has not been formally reviewed yet :) @yfarjoun's comment was just a one-off made while we were looking at the branch during our regular htsjdk review meeting this week. I'll do a full review now.

droazen commented on src/main/java/htsjdk/samtools/SamFiles.java:
```diff
@@ -19,61 +21,73 @@
+     *
+     * @return The index for the provided SAM, or null if one was not found.
+     */
+    public static Path findIndex(final Path samPath) {
```
`findIndex(Path)` needs direct unit testing -- one test case with a regular bai index file represented as a `Path`, and another with a symlink to a bai, so that we can confirm the existing functionality has not been broken by switching to `Path` here.

droazen commented on src/main/java/htsjdk/samtools/SamFiles.java:
```diff
@@ -19,61 +21,73 @@
-        final String fileName = samFile.getName();
+    private static Path lookForIndex(final Path samPath) {// If input is foo.bam, look for foo.bai
+        Path indexPath;
+        final String fileName = samPath.getFileName().toString();
```
Will this work if `samPath` is not a normal file (eg., is an HDFS URI or GCS URI?)

tomwhite replied:
Yes, it will. `samPath.getFileName()` returns the last component of the Path, which is then converted to a string.

droazen replied:
Can you add a comment here documenting that this will work correctly with URIs?

droazen commented on src/main/java/htsjdk/samtools/SamFiles.java:
```diff
@@ -19,61 +21,73 @@
-            if (indexFile.isFile()) {
-                return indexFile;
+            indexPath = samPath.resolveSibling(bai);
+            if (Files.isRegularFile(indexPath)) {
```
Same question here -- what happens if `indexPath` is a non-file URI?

tomwhite replied:
I just tried it with an HDFS Path (in the jsr-203 project) and it works as expected.

droazen commented on src/main/java/htsjdk/samtools/SamFiles.java:
```diff
@@ -19,61 +21,73 @@
+        indexPath = samPath.resolveSibling(fileName + BAMIndex.BAMIndexSuffix);
+        if (Files.isRegularFile(indexPath)) {
+            return indexPath;
         }
```
Add comment that this code is here and not in the `if (fileName.endsWith(BamFileIoUtils.BAM_FILE_EXTENSION)) {` block because we want to fall back to looking for bai's on cram files as well (at least, that is my interpretation of why this is here instead of above)

tomwhite replied:
The first block looks for foo.bai, whereas this one looks for foo.bam.bai. I think the comments cover this, but let me know if there's an improvement you'd like to see.

droazen replied:
The improvement I'd like to see is to mention that this fallback code at the bottom looks for bai indices on cram files as well (which is allowed). Eg., `look for foo.bam.bai or foo.cram.bai`

droazen commented on src/main/java/htsjdk/samtools/SamFiles.java:
```diff
@@ -19,61 +21,73 @@
-            final File canonicalIndexFile = lookForIndex(canonicalSamFile);
-            if ( canonicalIndexFile != null) {
-                System.err.println("The index file " + canonicalIndexFile.getPath()
+            final Path canonicalSamPath = samPath.toRealPath();
```
Add comment that `toRealPath()` resolves any symbolic links

droazen commented on src/main/java/htsjdk/samtools/SamInputResource.java:
```diff
@@ -147,6 +164,9 @@ public String toString() {
                 childToString = asFile().toString();
                 break;
+            case PATH:
+                childToString = asPath().toString();
```
Guard against null here (since `asPath()` can return null)

tomwhite replied:
This is an existing issue with all the other resource types, so I've left it to be consistent. This should really be fixed separately.

droazen commented on src/main/java/htsjdk/samtools/SamInputResource.java:
```diff
@@ -192,6 +212,62 @@ public File asFile() {
+
+    @Override
+    public URL asUrl() {
+        return null;
```
It's strange that `FileInputResource`s can be converted to `Path` but not `URL`, since `Path` itself is convertible to `URL` -- do you want to address this now?

tomwhite replied:
OK.

droazen commented on src/main/java/htsjdk/samtools/SamInputResource.java:
```diff
@@ -192,6 +212,62 @@ public File asFile() {
+
+    @Override
+    public File asFile() {
+        return null;
```
Should you attempt `Path.toFile()` here?

tomwhite replied:
Yes, done.

droazen commented on src/main/java/htsjdk/samtools/SamInputResource.java:
```diff
@@ -192,6 +212,62 @@ public File asFile() {
+
+    @Override
     public URL asUrl() {
         return null;
```
Call `Path.toUri().toURL()` here?

droazen commented on src/main/java/htsjdk/samtools/SamReaderFactory.java:
```diff
@@ -57,6 +58,13 @@
     
     abstract public SamReader open(final File file);
 
+    public SamReader open(final Path path) {
```
Should live in `SamReaderFactoryImpl` along with the other `open()` implementations

tomwhite replied:
The reason I didn't add an abstract method here was to maintain backwards compatibility. Is there a better way of adding the overloaded method?

droazen replied:
Ok, fair enough

droazen commented on src/main/java/htsjdk/samtools/SamReaderFactory.java:
```diff
@@ -57,6 +58,13 @@
+        final Path indexMaybe = SamFiles.findIndex(path);
+        if (indexMaybe != null) r.index(indexMaybe);
+        return open(r);
+    }
```
What if the `Path` is a cram, and the client wants to call `referenceSequence()` on the factory? Do we have a plan to make that work?

tomwhite replied:
That should work fine as the implementation of `open` will use the reference sequence set on the factory (if any).

droazen replied:
What I mean, though, is that `referenceSequence()` takes a `File` -- there is no overload that takes a `Path`. If you're not going to address this here, can you open a ticket for it?

tomwhite replied:
You can construct a `ReferenceSource` with a `Path` and pass that to `referenceSequence()` - so there is a way of doing it a bit less directly.

droazen commented on src/main/java/htsjdk/samtools/seekablestream/SeekablePathStream.java:
```diff
@@ -0,0 +1,108 @@
+        return position() - initPos;
+    }
+    
+    public int read(final byte[] buffer, final int offset, final int length) throws IOException {
```
`@Override` here and for all other implementations of methods from `SeekableStream` in this class

droazen commented on src/main/java/htsjdk/samtools/SamFiles.java:
```diff
@@ -19,61 +21,73 @@
+            final Path canonicalSamPath = samPath.toRealPath();
+            final Path canonicalIndexPath = lookForIndex(canonicalSamPath);
+            if ( canonicalIndexPath != null) {
+                System.err.println("The index file " + canonicalIndexPath.toAbsolutePath()
```
Gross! Could you change this to print the warning to the logger instead?

droazen commented on src/main/java/htsjdk/samtools/seekablestream/SeekablePathStream.java:
```diff
@@ -0,0 +1,108 @@
+    private final long length;
+    private final ByteBuffer oneByteBuf = ByteBuffer.allocate(1);
+
+    public SeekablePathStream(final Path path) throws IOException {
```
Ensure `path` is non-null

tomwhite replied:
I didn't do this because the other `Seekable*Stream` implementations don't.

droazen replied:
Ok

droazen commented on src/main/java/htsjdk/samtools/seekablestream/SeekablePathStream.java:
```diff
@@ -0,0 +1,108 @@
+    }
+
+    public long length() {
+        return length;
```
Could the length change between construction time and a call to `length()`?

tomwhite replied:
It could. The reason I didn't do this before was because `SeekableByteChannel.size()` can throw an `IOException`, but `SeekableStream.length` cannot. I've caught the exception, logged it, and returned 0 now (which is consistent with `java.io.File`).

droazen commented on src/main/java/htsjdk/samtools/seekablestream/SeekablePathStream.java:
```diff
@@ -0,0 +1,108 @@
+        return length() == position();
+    }
+
+    public void seek(final long position) throws IOException {
```
What do you expect seeking performance to be like if the `Path` is something like an HDFS file?

tomwhite replied:
Seeks should be avoided if possible, since it's better to stream through a file in HDFS. Having said that, in HDFS seeks are implemented pretty efficiently - as a regular file seek if within the same HDFS block, or as a transparent seek to another datanode (if the seek takes you to another block).

droazen commented on src/main/java/htsjdk/samtools/seekablestream/SeekablePathStream.java:
```diff
@@ -0,0 +1,108 @@
+    @Override
+    public long skip(long n) throws IOException {
+        long initPos = position();
+        sbc.position(initPos + n);
```
What is the behavior here if this brings us past `eof()`?

tomwhite replied:
It will succeed, but then any attempt to read will get an EOF. This is the same as `SeekableFileStream`, and is specified here: https://docs.oracle.com/javase/8/docs/api/java/nio/channels/SeekableByteChannel.html#position-long-

droazen commented on src/main/java/htsjdk/samtools/seekablestream/SeekablePathStream.java:
```diff
@@ -0,0 +1,108 @@
+     * duration of a session.  This collection supports a method to close them all.
+     */
+    static Collection<SeekablePathStream> allInstances =
+            Collections.synchronizedCollection(new HashSet<SeekablePathStream>());
```
Can this be private?

tomwhite replied:
Yes, done.

droazen commented on src/main/java/htsjdk/samtools/seekablestream/SeekablePathStream.java:
```diff
@@ -0,0 +1,108 @@
+import java.util.Collections;
+import java.util.HashSet;
+
+public class SeekablePathStream extends SeekableStream {
```
Add direct unit tests for this class that fully exercise `read()` (both versions), `skip()`, `seek()`, `position()`, and `length()` (can use a `Path` backed by a regular file for the tests).

droazen commented on src/test/java/htsjdk/samtools/SamReaderFactoryTest.java:
```diff
@@ -214,6 +217,24 @@ public void exhaustInputResourcePermutation(final SamInputResource resource) thr
+        observedHeaders.add(fileHeader);
+        observedRecordOrdering.add(slurped);
+        Assert.assertEquals(observedHeaders.size(), 1, "read different headers than other testcases");
+        Assert.assertEquals(observedRecordOrdering.size(), 1, "read different records than other testcases");
```
This test is not meaningful if it happens to run *before* `exhaustInputResourcePermutation()` has populated  `observedHeaders` and `observedRecordOrdering` instead of after it. This is a problem since TestNG by default randomizes test ordering.

droazen replied:
Recommend that instead of using `observedHeaders` and `observedRecordOrdering` directly in this test case, you build the expected values separately within this method using `SamReaderFactory.open(file)`, and then assert that the results you get from `SamReaderFactory.open(path)` match the ones you got from `open(file)`. That way you don't share data with `exhaustInputResourcePermutation()` and don't have to worry about test ordering issues.

droazen commented:
Review complete -- back to @tomwhite for changes.

tomwhite commented:
Thanks for the review @droazen. I've addressed all the points you made - a few (which I've addressed in a comment) are longstanding issues that would be best addressed separately. Back to you.

droazen commented on src/test/java/htsjdk/samtools/SamFilesTest.java:
```diff
@@ -58,6 +60,15 @@ public void testFindIndexForSuffixes(final String dataFileSuffix, final String i
+        final Path foundIndexPath = SamFiles.findIndex(dataFile.toPath());
+        if (expectIndexSuffix == null) {
+            Assert.assertNull(foundIndexPath);
+            return;
```
Refactor out these `return` statements from this test case, replacing with `else` blocks -- otherwise when `expectIndexSuffix == null` you'll only test the `File` case and not the `Path` case

droazen commented on src/main/java/htsjdk/samtools/SamFiles.java:
```diff
@@ -19,61 +24,73 @@
+            final Path canonicalSamPath = samPath.toRealPath(); // resolve symbolic links
+            final Path canonicalIndexPath = lookForIndex(canonicalSamPath);
+            if ( canonicalIndexPath != null) {
+                LOG.error("The index file " + canonicalIndexPath.toAbsolutePath()
```
I think this should be a `LOG.warn()`, not `LOG.error()`

droazen commented on src/main/java/htsjdk/samtools/seekablestream/SeekablePathStream.java:
```diff
@@ -0,0 +1,118 @@
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+
```
Add class-level javadoc for this new class

droazen commented on src/main/java/htsjdk/samtools/seekablestream/SeekablePathStream.java:
```diff
@@ -0,0 +1,118 @@
+     * duration of a session.  This collection supports a method to close them all.
+     */
+    private static Collection<SeekablePathStream> allInstances =
+            Collections.synchronizedCollection(new HashSet<SeekablePathStream>());
```
Can this be final?

droazen commented on src/test/java/htsjdk/samtools/SamReaderFactoryTest.java:
```diff
@@ -214,6 +217,24 @@ public void exhaustInputResourcePermutation(final SamInputResource resource) thr
+    @Test
+    public void openPath() throws IOException {
+        final Path path = localBam.toPath();
+        final SamReader reader = SamReaderFactory.makeDefault().open(path);
```
Close both readers in this test case when done using try-with-resources

droazen commented on src/test/java/htsjdk/samtools/SamReaderFactoryTest.java:
```diff
@@ -214,6 +217,24 @@ public void exhaustInputResourcePermutation(final SamInputResource resource) thr
+        LOG.info(String.format("Reading from %s ...", path));
+        final List<SAMRecord> records = Iterables.slurp(reader);
+        final SAMFileHeader fileHeader = reader.getFileHeader();
+        reader.hasIndex();;
```
Do you want to assert that `hasIndex()` is true? Also, remove stray semicolon.

droazen commented:
Second-pass review complete, back to @tomwhite. Please squash/rebase after addressing remaining comments so we can merge.

tomwhite commented:
Thanks for the second review @droazen. I've addressed all the remaining comments.

droazen commented:
:+1: merging!

tomwhite commented:
Thanks!
